File: te_phoenix/lib/te_phoenix/battle/stats.ex
defmodule TePhoenix.Battle.Stats do
  require Logger
  @moduledoc """
  Effective stat calculator — loads a character from DB and builds
  a complete combatant stats map with equipment bonuses, status mods,
  ogham bonuses, reaction skills, and weapon types.
  Ported from te/battle/stats.js getEffectiveStats()
  """

  alias TePhoenix.Repo
  import Ecto.Query

  @doc """
  Load a character's effective combat stats from the database.
  Applies: base stats + equipment bonuses + status mods + ogham bonuses.
  Returns a map ready to be used as Combatant init data, or nil if not found.
  """
  def get_effective_stats(char_id) do
    # 1. Base character stats
    case Repo.one(from c in "characters", where: c.id == ^char_id,
           select: %{
             id: c.id, user_id: c.user_id, name: c.name, level: c.level,
             class_id: c.class_id, race_id: c.race_id,
             current_hp: c.current_hp, max_hp: c.max_hp,
             current_mp: c.current_mp, max_mp: c.max_mp,
             atk: c.atk, def: c.def, mo: c.mo, md: c.md,
             speed: c.speed, luck: c.luck,
             limitbreak: c.limitbreak, breaklevel: c.breaklevel,
             status_effects: c.status_effects, map_id: c.map_id,
             x: c.x, y: c.y, experience: c.experience
           }) do
      nil -> nil
      c ->
        stats = %{
          char_id: c.id,
          user_id: c.user_id,
          name: c.name,
          level: c.level || 1,
          class_id: c.class_id,
          race_id: c.race_id,
          current_hp: c.current_hp,
          max_hp: c.max_hp,
          current_mp: c.current_mp,
          max_mp: c.max_mp,
          atk: c.atk,
          def: c.def,
          mo: c.mo,
          md: c.md,
          speed: c.speed,
          luck: c.luck,
          limitbreak: parse_float(c.limitbreak) || 0,
          breaklevel: c.breaklevel || 1,
          statuses: parse_json_list(c.status_effects),
          weapon_elements: [],
          weapon_statuses: %{},
          elem_defenses: [],
          ogham_elements: [],
          ogham_statuses: %{},
          reactions: [],
          weapon_type: nil,
          armor_type: nil,
          penetration: 0,
          crit_resist: 0,
          map_id: c.map_id,
          experience: c.experience || 0
        }

        stats
        |> apply_equipment_bonuses(char_id)
        |> apply_status_mods()
        |> apply_ogham_bonuses(char_id)
        |> load_reactions(c.class_id, c.level)
        |> detect_weapon_type(char_id)
    end
  rescue
    e ->
      require Logger
      Logger.warning("Failed to load stats for char #{char_id}: #{inspect(e)}")
      nil
  end

  # ── Equipment bonuses ───────────────────────────────────────────

  defp apply_equipment_bonuses(stats, char_id) do
    equip =
      try do
        Repo.all(
          from ce in "character_equipment",
          join: gi in "game_items", on: ce.item_id == gi.id,
          where: ce.character_id == ^char_id,
          select: %{
            slot_key: ce.slot_key, item_id: gi.id,
            bonus_hp: gi.bonus_hp, bonus_mp: gi.bonus_mp,
            bonus_atk: gi.bonus_atk, bonus_def: gi.bonus_def,
            bonus_mo: gi.bonus_mo, bonus_md: gi.bonus_md,
            bonus_speed: gi.bonus_speed, bonus_luck: gi.bonus_luck,
            elements: gi.elements, set_status: gi.set_status,
            type: gi.type
          }
        )
      rescue
        _ -> []
      end

    Enum.reduce(equip, stats, fn item, acc ->
      acc = %{acc |
        atk: acc.atk + (item.bonus_atk || 0),
        def: acc.def + (item.bonus_def || 0),
        mo: acc.mo + (item.bonus_mo || 0),
        md: acc.md + (item.bonus_md || 0),
        speed: acc.speed + (item.bonus_speed || 0),
        luck: acc.luck + (item.bonus_luck || 0),
        max_hp: acc.max_hp + (item.bonus_hp || 0),
        max_mp: acc.max_mp + (item.bonus_mp || 0)
      }

      # Element processing from equipment
      acc = case parse_json_map(item.elements) do
        nil -> acc
        elems ->
          Enum.reduce(elems, acc, fn {elem_name, val}, a ->
            en = String.downcase(to_string(elem_name))
            case val do
              %{"role" => "attack"} ->
                %{a | weapon_elements: [en | a.weapon_elements]}
              %{"role" => role, "pct" => pct} ->
                %{a | elem_defenses: [%{elem: en, role: role, pct: pct} | a.elem_defenses]}
              %{"role" => role} ->
                %{a | elem_defenses: [%{elem: en, role: role, pct: 50} | a.elem_defenses]}
              "attack" ->
                %{a | weapon_elements: [en | a.weapon_elements]}
              _ ->
                %{a | elem_defenses: [%{elem: en, role: "resist", pct: 50} | a.elem_defenses]}
            end
          end)
      end

      # Weapon status effects (from main/off hand)
      acc = if item.slot_key in ["MAIN_HAND", "OFF_HAND"] do
        case parse_json_map(item.set_status) do
          nil -> acc
          ws -> %{acc | weapon_statuses: Map.merge(acc.weapon_statuses, ws)}
        end
      else
        acc
      end

      acc
    end)
  end

  # ── Status effect modifiers ─────────────────────────────────────

  defp apply_status_mods(stats) do
    Enum.reduce(stats.statuses, stats, fn status, acc ->
      status_id = status["id"] || status[:id]
      if status_id == nil, do: acc

      case load_status_effects(status_id) do
        nil -> acc
        effects ->
          # Apply stat_mod multipliers
          case Map.get(effects, "stat_mod") do
            nil -> acc
            mods ->
              Enum.reduce(mods, acc, fn {stat_key, multiplier}, a ->
                try do
                  key = String.to_existing_atom(stat_key)
                  case Map.get(a, key) do
                    val when is_number(val) -> Map.put(a, key, trunc(val * multiplier))
                    _ -> a
                  end
                rescue
                  _ -> a
                end
              end)
          end
      end
    end)
  end

  defp load_status_effects(status_id) do
    case Repo.one(from s in "game_statuses", where: s.id == ^status_id, select: s.effects) do
      nil -> nil
      json -> parse_json_map(json)
    end
  rescue
    _ -> nil
  end

  # ── Ogham bonuses ───────────────────────────────────────────────

  defp apply_ogham_bonuses(stats, char_id) do
    oghams =
      try do
        Repo.all(
          from co in "character_oghams",
          join: go in "game_oghams", on: go.id == co.ogham_id,
          where: co.character_id == ^char_id,
          select: %{
            element_attack: go.element_attack,
            on_hit_status: go.on_hit_status,
            on_hit_chance: go.on_hit_chance,
            stat_bonus_json: go.stat_bonus_json
          }
        )
      rescue
        _ -> []
      end

    Enum.reduce(oghams, stats, fn og, acc ->
      # Element attack
      acc = if og.element_attack do
        %{acc | ogham_elements: [String.downcase(og.element_attack) | acc.ogham_elements]}
      else
        acc
      end

      # On-hit status
      acc = if og.on_hit_status do
        %{acc | ogham_statuses: Map.put(acc.ogham_statuses, og.on_hit_status, %{
          chance: og.on_hit_chance || 20
        })}
      else
        acc
      end

      # Stat bonuses
      case parse_json_map(og.stat_bonus_json) do
        nil -> acc
        bonuses ->
          Enum.reduce(bonuses, acc, fn {k, v}, a ->
            try do
              key = String.to_existing_atom(k)
              case Map.get(a, key) do
                val when is_number(val) -> Map.put(a, key, val + v)
                _ -> a
              end
            rescue
              _ -> a
            end
          end)
      end
    end)
  end

  # ── Reaction skills ─────────────────────────────────────────────

  defp load_reactions(stats, class_id, level) do
    reactions =
      try do
        Repo.all(
          from gcs in "game_class_skills",
          join: gs in "game_skills", on: gcs.skill_id == gs.id,
          where: gcs.class_id == ^class_id and gcs.learn_level <= ^level,
          select: %{id: gs.id, name: gs.name, icon: gs.icon, effects: gs.effects}
        )
        |> Enum.filter(fn s ->
          case parse_json_map(s.effects) do
            %{"reaction" => %{"trigger" => _}} -> true
            _ -> false
          end
        end)
        |> Enum.map(fn s ->
          fx = parse_json_map(s.effects)
          %{
            skill_id: s.id,
            name: s.name,
            icon: s.icon,
            trigger: get_in(fx, ["reaction", "trigger"]),
            chance: get_in(fx, ["reaction", "chance"]) || 25
          }
        end)
      rescue
        _ -> []
      end

    %{stats | reactions: reactions}
  end

  # ── Weapon type detection ───────────────────────────────────────

  defp detect_weapon_type(stats, char_id) do
    weapon_type =
      try do
        Repo.one(
          from ce in "character_equipment",
          join: gi in "game_items", on: ce.item_id == gi.id,
          where: ce.character_id == ^char_id and ce.slot_key == "MAIN_HAND",
          select: gi.type
        )
      rescue
        _ -> nil
      end

    %{stats | weapon_type: weapon_type}
  end

  # ── Enemy scaling for party battles ─────────────────────────────

  @doc "Scale enemy stats based on player count"
  def apply_enemy_scaling(enemy_stats, player_count, scaling_factor \\ 1.0)

  def apply_enemy_scaling(enemy_stats, player_count, scaling_factor) when player_count > 1 do
    # Each additional player adds ~40% HP and ~15% stats
    hp_mult = 1.0 + (player_count - 1) * 0.40 * scaling_factor
    stat_mult = 1.0 + (player_count - 1) * 0.15 * scaling_factor

    %{enemy_stats |
      max_hp: trunc(enemy_stats.max_hp * hp_mult),
      current_hp: trunc(enemy_stats.current_hp * hp_mult),
      atk: trunc(enemy_stats.atk * stat_mult),
      def: trunc(enemy_stats.def * stat_mult),
      mo: trunc(enemy_stats.mo * stat_mult),
      md: trunc(enemy_stats.md * stat_mult)
    }
  end

  def apply_enemy_scaling(stats, _, _), do: stats

  # ── JSON helpers ────────────────────────────────────────────────

  defp parse_json_map(nil), do: nil
  defp parse_json_map(val) when is_map(val), do: val
  defp parse_json_map(str) when is_binary(str) do
    case Jason.decode(str) do
      {:ok, map} when is_map(map) -> map
      _ -> nil
    end
  end
  defp parse_json_map(_), do: nil

  defp parse_json_list(nil), do: []
  defp parse_json_list(val) when is_list(val), do: val
  defp parse_json_list(str) when is_binary(str) do
    case Jason.decode(str) do
      {:ok, list} when is_list(list) -> list
      _ -> []
    end
  end
  defp parse_json_list(_), do: []

  defp parse_float(nil), do: 0.0
  defp parse_float(f) when is_float(f), do: f
  defp parse_float(i) when is_integer(i), do: i * 1.0
  defp parse_float(s) when is_binary(s) do
    case Float.parse(s) do
      {f, _} -> f
      :error -> 0.0
    end
  end
  defp parse_float(_), do: 0.0
end

File: te.retired/battle/stats.js
// =================================================================
// BATTLE STATS — Effective stat calculator, formula vars, scaling
// =================================================================

const { safeEval } = require('../event_runner');

function jp(s, f) { try { return JSON.parse(s); } catch { return f; } }

// =================================================================
// TABLE COMPATIBILITY HELPERS
// =================================================================
async function queryLevelRow(db, level) {
    for (const tbl of ['level_requirements', 'game_levels']) {
        try {
            const [rows] = await db.query(`SELECT * FROM \`${tbl}\` WHERE level=?`, [level]);
            return rows;
        } catch (e) {
            if (e.code === 'ER_NO_SUCH_TABLE' || e.errno === 1146) continue;
            throw e;
        }
    }
    return [];
}

async function queryLimitBreakRow(db, id, classId) {
    for (const tbl of ['game_limit_breaks', 'game_limits']) {
        try {
            const [rows] = await db.query(`SELECT * FROM \`${tbl}\` WHERE id=? AND class_id=?`, [id, classId]);
            return rows;
        } catch (e) {
            if (e.code === 'ER_NO_SUCH_TABLE' || e.errno === 1146) continue;
            throw e;
        }
    }
    return [];
}

async function queryLimitBreaksList(db, classId, charLevel, breakLevel) {
    for (const tbl of ['game_limit_breaks', 'game_limits']) {
        try {
            const [rows] = await db.query(
                `SELECT * FROM \`${tbl}\` WHERE class_id=? AND char_level_req<=? AND break_level<=? ORDER BY break_level`,
                [classId, charLevel, breakLevel]
            );
            return rows;
        } catch (e) {
            if (e.code === 'ER_NO_SUCH_TABLE' || e.errno === 1146) continue;
            throw e;
        }
    }
    return [];
}

// =================================================================
// STAT CALCULATOR — Equipment + Status Modifiers
// =================================================================
async function getEffectiveStats(db, charId) {
    // 1. Base stats
    const [charRows] = await db.query("SELECT * FROM characters WHERE id=?", [charId]);
    if (!charRows.length) return null;
    const c = charRows[0];

    const stats = {
        charId: c.id,
        userId: c.user_id,
        name: c.name,
        level: c.level,
        classId: c.class_id,
        raceId: c.race_id,
        currentHp: c.current_hp,
        maxHp: c.max_hp,
        currentMp: c.current_mp,
        maxMp: c.max_mp,
        atk: c.atk,
        def: c.def,
        mo: c.mo,
        md: c.md,
        speed: c.speed,
        luck: c.luck,
        limitbreak: parseFloat(c.limitbreak) || 0,
        breaklevel: c.breaklevel || 1,
        statuses: jp(c.status_effects, []),
        weaponElements: [],
        weaponStatuses: {},
        armorBlockStatuses: [],
        elemDefenses: [],
        _penetration: 0,
        _critResist: 0,
        oghamElements: [],
        oghamStatuses: {},
        experience: c.experience || 0
    };

    // 2. Equipment bonuses
    const [equip] = await db.query(`
        SELECT ce.slot_key, gi.* FROM character_equipment ce
        JOIN game_items gi ON ce.item_id = gi.id
        WHERE ce.character_id = ?`, [charId]);

    for (const item of equip) {
        stats.atk += (item.bonus_atk || 0);
        stats.def += (item.bonus_def || 0);
        stats.mo  += (item.bonus_mo || 0);
        stats.md  += (item.bonus_md || 0);
        stats.speed += (item.bonus_speed || 0);
        stats.luck  += (item.bonus_luck || 0);
        stats.maxHp += (item.bonus_hp || 0);
        stats.maxMp += (item.bonus_mp || 0);

        const itemExtras = jp(item.combat_bonuses, null);
        if (itemExtras) {
            if (itemExtras.penetration) stats._penetration += parseFloat(itemExtras.penetration) || 0;
            if (itemExtras.crit_resist) stats._critResist += parseFloat(itemExtras.crit_resist) || 0;
        }

        const elems = jp(item.elements, null);
        if (elems) {
            for (const [elemName, val] of Object.entries(elems)) {
                const en = elemName.toLowerCase();
                if (typeof val === 'object' && val !== null) {
                    const role = val.role || 'resist';
                    const pct  = val.pct !== undefined ? val.pct : 50;
                    if (role === 'attack') {
                        stats.weaponElements.push(en);
                    } else {
                        stats.elemDefenses.push({ elem: en, role, pct });
                    }
                } else {
                    if (val === 'attack') stats.weaponElements.push(en);
                    else stats.elemDefenses.push({ elem: en, role: 'resist', pct: 50 });
                }
            }
        }

        if (item.slot_key === 'MAIN_HAND' || item.slot_key === 'OFF_HAND') {
            const ws = jp(item.set_status, null);
            if (ws) Object.assign(stats.weaponStatuses, ws);
        }

        const blocked = jp(item.block_status, null);
        if (blocked) stats.armorBlockStatuses.push(...blocked);
    }

    // 3. Status effect modifiers
    for (const status of stats.statuses) {
        const [sRows] = await db.query("SELECT * FROM game_statuses WHERE id=?", [status.id]);
        if (!sRows.length) continue;
        const effects = jp(sRows[0].effects, {});
        if (effects.stat_mod) {
            for (const [statKey, multiplier] of Object.entries(effects.stat_mod)) {
                if (stats[statKey] !== undefined && typeof stats[statKey] === 'number') {
                    stats[statKey] = Math.floor(stats[statKey] * multiplier);
                }
            }
        }
        // Aggregate special combat modifiers from status effects
        if (effects.element_resist) {
            if (!stats._elementResist) stats._elementResist = {};
            for (const [elem, mult] of Object.entries(effects.element_resist)) {
                stats._elementResist[elem] = (stats._elementResist[elem] || 1) * mult;
            }
        }
        if (effects.element_infuse) {
            if (!stats._elementInfuse) stats._elementInfuse = [];
            stats._elementInfuse.push(effects.element_infuse);
        }
        if (effects.thorns) stats._thorns = (stats._thorns || 0) + effects.thorns;
        if (effects.counter_chance) stats._counterChance = Math.min(100, (stats._counterChance || 0) + effects.counter_chance);
        if (effects.crit_chance_mod) stats._critChanceMod = (stats._critChanceMod || 0) + effects.crit_chance_mod;
        if (effects.evasion_mod) stats._evasionMod = (stats._evasionMod || 0) + effects.evasion_mod;
    }

    // 4. Blood Ogham bonuses
    const [oghams] = await db.query(`
        SELECT co.*, go.element_attack, go.on_hit_status, go.on_hit_chance,
               go.stat_bonus_json, go.grant_skill_id, go.curse_json, go.family_id
        FROM character_oghams co
        JOIN game_oghams go ON go.id = co.ogham_id
        WHERE co.character_id = ?`, [charId]);

    for (const og of oghams) {
        if (og.element_attack) stats.oghamElements.push(og.element_attack.toLowerCase());
        if (og.on_hit_status) {
            stats.oghamStatuses[og.on_hit_status] = {
                chance: og.on_hit_chance || 20,
                curse:  jp(og.curse_json, null)
            };
        }
        const bonus = jp(og.stat_bonus_json, null);
        if (bonus) {
            for (const [k, v] of Object.entries(bonus)) {
                if (typeof stats[k] === 'number') stats[k] += v;
            }
        }
    }

    // Set bonus: 2+ Oghams from same family
    const familyCounts = {};
    for (const og of oghams) {
        if (og.family_id) familyCounts[og.family_id] = (familyCounts[og.family_id] || 0) + 1;
    }
    const activeFamily = Object.keys(familyCounts).filter(fid => familyCounts[fid] >= 2);
    if (activeFamily.length) {
        try {
            const [families] = await db.query(
                'SELECT * FROM game_ogham_families WHERE id IN (?)', [activeFamily]);
            for (const fam of families) {
                const sb = jp(fam.set_bonus_json, null);
                if (!sb) continue;
                if (sb.stat_bonus) {
                    for (const [k, v] of Object.entries(sb.stat_bonus)) {
                        if (typeof stats[k] === 'number') stats[k] += v;
                    }
                }
                if (sb.element_attack) stats.oghamElements.push(sb.element_attack.toLowerCase());
                if (sb.on_hit_status) {
                    stats.oghamStatuses[sb.on_hit_status] = {
                        chance: sb.on_hit_chance || 20, curse: null
                    };
                }
                if (!stats.activeSets) stats.activeSets = [];
                stats.activeSets.push({ name: fam.name, label: sb.label || fam.name });
            }
        } catch(setErr) { console.warn('Set bonus error (non-fatal):', setErr.message); }
    }

    // 5. Passive reaction skills
    try {
        const [reactRows] = await db.query(`
            SELECT gs.id, gs.name, gs.icon, gs.effects, gcs.learn_level
            FROM game_class_skills gcs
            JOIN game_skills gs ON gcs.skill_id = gs.id
            WHERE gcs.class_id = ? AND gcs.learn_level <= ?`, [stats.classId, stats.level]);
        stats.reactions = [];
        for (const r of reactRows) {
            const rfx = jp(r.effects, {});
            if (rfx.reaction && rfx.reaction.trigger) {
                stats.reactions.push({
                    skillId: r.id,
                    name:    r.name,
                    icon:    r.icon,
                    trigger: rfx.reaction.trigger,
                    chance:  rfx.reaction.chance || 25
                });
            }
        }
    } catch { stats.reactions = []; }

    // 6. Body type (for limb targeting)
    stats.bodyTypeId = 1;
    try {
        const [npcRow] = await db.query(
            'SELECT body_type_id FROM game_npcs WHERE char_id=? LIMIT 1', [charId]);
        if (npcRow.length && npcRow[0].body_type_id) {
            stats.bodyTypeId = npcRow[0].body_type_id;
        }
    } catch {}

    // Merge ogham elements into weapon elements
    stats.weaponElements = [...new Set([...stats.weaponElements, ...stats.oghamElements])];

    // Clamp HP to max
    if (stats.currentHp > stats.maxHp) stats.currentHp = stats.maxHp;
    if (stats.currentMp > stats.maxMp) stats.currentMp = stats.maxMp;

    return stats;
}

// Build the vars object for safeEval formulas
function buildFormulaVars(attacker, defender) {
    return {
        ATK:        attacker.atk,
        DEF:        defender.def,
        MO:         attacker.mo,
        MD:         defender.md,
        SPEED:      attacker.speed,
        LUCK:       attacker.luck,
        HP:         attacker.currentHp,
        MP:         attacker.currentMp,
        MAXHP:      attacker.maxHp,
        MAXMP:      attacker.maxMp,
        HP_PCT:     attacker.maxHp > 0 ? Math.floor((attacker.currentHp / attacker.maxHp) * 100) : 0,
        LIMITBREAK: attacker.limitbreak || 0,
        LVL:        attacker.level,
        ENEMY_ATK:      defender.atk,
        ENEMY_DEF:      defender.def,
        ENEMY_MO:       defender.mo,
        ENEMY_MD:       defender.md,
        ENEMY_SPEED:    defender.speed,
        ENEMY_LUCK:     defender.luck,
        ENEMY_HP:       defender.currentHp,
        ENEMY_MAXHP:    defender.maxHp,
        ENEMY_HP_PCT:   defender.maxHp > 0 ? Math.floor((defender.currentHp / defender.maxHp) * 100) : 0,
        ENEMY_LVL:      defender.level
    };
}

// =================================================================
// ENEMY SCALING
// =================================================================
function applyEnemyScaling(enemyStats, playerCount, scalingFactor = 0.3) {
    if (playerCount <= 1 || scalingFactor <= 0) return enemyStats;
    const mult = 1 + (playerCount - 1) * scalingFactor;
    const scaled = { ...enemyStats };
    scaled.maxHp     = Math.round(scaled.maxHp * mult);
    scaled.currentHp = Math.round(scaled.currentHp * mult);
    scaled.atk       = Math.round(scaled.atk * mult);
    scaled.def       = Math.round(scaled.def * mult);
    scaled.mo        = Math.round(scaled.mo * mult);
    scaled.md        = Math.round(scaled.md * mult);
    scaled.speed     = Math.round(scaled.speed * (1 + (playerCount - 1) * scalingFactor * 0.3));
    return scaled;
}

async function getScalingFactor(db, mapId) {
    try {
        const [zones] = await db.query(
            'SELECT scaling_factor FROM game_map_spawns WHERE map_id=? AND enabled=1 AND scaling_factor > 0 LIMIT 1',
            [mapId]);
        if (zones.length) return parseFloat(zones[0].scaling_factor) || 0.3;
    } catch {}
    try {
        const [rows] = await db.query(
            "SELECT setting_value FROM system_settings WHERE setting_key='enemy_scaling_factor' LIMIT 1");
        if (rows.length) return parseFloat(rows[0].setting_value) || 0.3;
    } catch {}
    return 0.3;
}

module.exports = {
    jp,
    getEffectiveStats,
    buildFormulaVars,
    queryLevelRow,
    queryLimitBreakRow,
    queryLimitBreaksList,
    applyEnemyScaling,
    getScalingFactor
};
